### PR TITLE
o, o/snapstate, tests: on classic auto-install snapd as prereq for non-essential snap install

### PR DESCRIPTION
**Pull request by ernestl**
On classic systems, auto install snapd as prerequisite for any snap that is not os, base, kernel, gadget or snapd.

Previously it was only installed on systems without core, leaving a small amount of systems with snapd deb + core - inconsistent with the majority of systems in terms of re-execution and upgrade path.

The change triggered an explosion of unit test failures. The most efficient approach for addressing this was to install snapd as part of the base test (prevents many breakages) and remove it where required (only a few places).
This mirrors the spread change: https://github.com/canonical/snapd/pull/14294 

Jira: [SNAPDENG-24011](https://warthogs.atlassian.net/browse/SNAPDENG-24011)

This change overlaps with existing experimental feature: https://github.com/canonical/snapd/blob/master/overlord/snapstate/snapmgr.go#L1167
We should consider deprecating this in a followup.

[SNAPDENG-24011]: https://warthogs.atlassian.net/browse/SNAPDENG-24011?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: overlord/snapstate/handlers.go [hunk 1]
```diff
@@ -383,11 +383,19 @@ func (m *SnapManager) installOneBaseOrRequired(t *state.Task, snapName string, c
 	// retrying the operation.
 	var conflErr *ChangeConflictError
 	if errors.As(err, &conflErr) {
+		// TODO: Fix unit test to enable this
+		// if we are not seeded it is too early to do installs
```
Comment (ernestl) [inline overlord/snapstate/handlers.go:387]: This proposal needs to be discussed

File: overlord/snapstate/autorefresh_gating_test.go [hunk 2]
```diff
@@ -1960,6 +1960,14 @@ func (s *snapmgrTestSuite) testAutoRefreshPhase2(c *C, beforePhase1 func(), gate
 	st.Lock()
 	defer st.Unlock()
 
+	// prevent automatic installation of snapd
```
Comment (ernestl) [inline overlord/snapstate/autorefresh_gating_test.go:1963]: Need to remove this

File: overlord/snapstate/autorefresh_gating_test.go [hunk 3]
```diff
@@ -2255,6 +2263,14 @@ func (s *snapmgrTestSuite) testAutoRefreshPhase2DiskSpaceCheck(c *C, fail bool)
...
```
Comment (ernestl) [inline overlord/snapstate/autorefresh_gating_test.go:2266]: Need to remove this

**Comment by ernestl** [comment]
FAILURES:

nested-ubuntu-22.04:   All are known, not caused by changes
 - google-nested:ubuntu-22.04-64:tests/nested/manual/muinstaller-core:install_optional_all
- google-nested:ubuntu-22.04-64:tests/nested/manual/muinstaller-core:install_optional_snap
- google-nested:ubuntu-22.04-64:tests/nested/manual/muinstaller-core:install_optional_snap_and_comp
- google-nested:ubuntu-22.04-64:tests/nested/manual/muinstaller-core:plain
- google-nested:ubuntu-22.04-64:tests/nested/manual/muinstaller-real:encrypted
- google-nested:ubuntu-22.04-64:tests/nested/manual/muinstaller-real:partial
- google-nested:ubuntu-22.04-64:tests/nested/manual/muinstaller-real:plain

nested-ubuntu-24.04:  All are known, not caused by changes
Failed:                    
- google-nested:ubuntu-24.04-64:tests/nested/core/core20-kernel-failover:crash
- google-nested:ubuntu-24.04-64:tests/nested/core/core20-kernel-reseal
- google-nested:ubuntu-24.04-64:tests/nested/manual/core20-early-config
- google-nested:ubuntu-24.04-64:tests/nested/manual/core20-initramfs-time-moves-forward
- google-nested:ubuntu-24.04-64:tests/nested/manual/muinstaller-core:install_optional_all
- google-nested:ubuntu-24.04-64:tests/nested/manual/muinstaller-core:install_optional_snap
- google-nested:ubuntu-24.04-64:tests/nested/manual/muinstaller-core:install_optional_snap_and_comp
- google-nested:ubuntu-24.04-64:tests/nested/manual/muinstaller-core:plain
- google-nested:ubuntu-24.04-64:tests/nested/manual/muinstaller-real:encrypted
- google-nested:ubuntu-24.04-64:tests/nested/manual/muinstaller-real:partial
- google-nested:ubuntu-24.04-64:tests/nested/manual/muinstaller-real:plain
- google-nested:ubuntu-24.04-64:tests/nested/manual/remodel-min-size
prepare:
- google-nested:ubuntu-24.04-64:tests/nested/manual/hybrid-remodel

File: tests/main/snapd-snap-transition-auto-install/task.yaml [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    else
+        echo "Ensure re-exec from core snap"
+        expect="DEBUG: restarting into \"$snap_mount_dir/core/current/usr/bin/snap\""
+    fi
+    SNAPD_DEBUG=1 snap list 2>&1 | MATCH "$expect"
+
+    # Install snaps with snaps with complex interface requirements.
+    # Use specific revisions to pin expected behaviour.
+    install_snaps() {
+        snap install firefox --revision=4848      # 130.0-2
```
Comment (bboozzoo) [inline tests/main/snapd-snap-transition-auto-install/task.yaml:60]: maybe drop firefox? it's large, has many interfaces and for sure will pull in additional content snap dependencies

File: overlord/snapstate/handlers.go [hunk 5]
```diff
@@ -560,18 +560,20 @@ func (m *SnapManager) installPrereqs(t *state.Task, base string, prereq map[stri
...
 	}
-	coreSnapInstalled, err := isInstalled(st, "core")
-	if err != nil {
-		return err
-	}
-	if base != "core" && !snapdSnapInstalled && !coreSnapInstalled {
+
+	// consider the state of seeding to avoid seed conflict error
+	deviceCtx, err := DevicePastSeeding(st, nil)
+	seeded := err == nil && deviceCtx != nil
```
Comment (pedronis) [inline overlord/snapstate/handlers.go:574]: we usually just get "seeded" from state to check for this

**Review by pedronis — COMMENTED** [review]
remark

File: tests/main/snapd-snap-transition-auto-install/task.yaml (hunk 4 above)
Comment (ernestl) [reply tests/main/snapd-snap-transition-auto-install/task.yaml:64]: Can do, if its ok, will wait a bit to gather more input on what constitutes a solid check for interface "migration".

File: tests/main/snapd-snap-transition-auto-install/task.yaml [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+    # 'deb' : transition snapd deb to re-exec from snapd snap
+    # 'core': transition re-exec from snapd in core snap to re-exec from snapd snap
+    if [ "$SNAPD_SRC" = "deb" ]; then
+        # core < deb, exec snapd deb
+        apt install -y --allow-downgrades snapd=2.63+22.04ubuntu0.1
+    else
+        # core => deb, re-exec snapd from core snap
+        apt install -y --allow-downgrades snapd=2.55.3+22.04
+    fi
+    snap install core --revision=17200 # 16-2.61.4-20240607
```
[the remark this answers is not in the document]
Comment (ernestl) [reply tests/main/snapd-snap-transition-auto-install/task.yaml:40]: Thanks, I clarified this

File: overlord/snapstate/handlers.go (hunk 5 above)
Comment (ernestl) [reply overlord/snapstate/handlers.go]: Changed this, also send question over MM for further consideration.

File: tests/main/snapd-snap-transition-auto-install/task.yaml [hunk 7]
```diff
@@ -38,6 +39,9 @@ prepare: |
     fi
     snap install core --revision=17200 # 16-2.61.4-20240607
 
+    # installation of core should not trigger prerequisites, but let's double check snapd is not installed
+    ! snap list snap
```
Comment (bboozzoo) [inline tests/main/snapd-snap-transition-auto-install/task.yaml:43]:
```suggestion
    ! snap list snapd
```

**Review by bboozzoo — APPROVED** [review]

**Review by pedronis — APPROVED** [review]
thanks

File: tests/main/snapd-snap-transition-auto-install/task.yaml (hunk 7 above)
Comment (ernestl) [reply tests/main/snapd-snap-transition-auto-install/task.yaml]: fixed :)

File: tests/main/snapd-snap-transition-auto-install/task.yaml [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+        expected_snap_with_snapd_prereq="hello-world"
+    fi
+
+    echo "Ensure no change errors during installation"
+    snap changes | tail -n +"$change_marker" | NOMATCH "Error"
+    test "$(snap changes | tail -n +"$change_marker" | grep -c "Done")" = "$expected_changes_after_mark"
+
+    echo "Ensure snapd snap was successfully automatically installed"
+    snap list snapd
+    change=$(snap changes | tail -n +"$change_marker" | grep "Install \"$expected_snap_with_snapd_prereq\"" | awk '{print $1}')
```
Comment (sergiocazzolato) [inline tests/main/snapd-snap-transition-auto-install/task.yaml:110]: is the tail needed?

File: tests/main/snapd-snap-transition-auto-install/task.yaml [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+# Simplify test by limiting system to ubuntu 22.04 amd64
+systems: [ubuntu-22.04-64]
+
+environment:
+    SNAPD_SRC/deb:  "deb"
+    SNAPD_SRC/core: "core"
+
+prepare: |
+    # remove all snaps to remove snapd
+    snap remove core
```
Comment (sergiocazzolato) [inline tests/main/snapd-snap-transition-auto-install/task.yaml:19]: I would check if it installed first, then I'll remove it

File: tests/main/snapd-snap-transition-auto-install/task.yaml [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+    # this is only possible when snapd is the only installed snap
+    snap remove snapd
```
Comment (sergiocazzolato) [inline tests/main/snapd-snap-transition-auto-install/task.yaml:21]: I would check if it installed first, then I'll remove it

File: tests/main/snapd-snap-transition-auto-install/task.yaml [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+        # core < deb, exec snapd deb
+        apt install -y --allow-downgrades snapd=2.63+22.04ubuntu0.1
+    else
+        # core => deb, re-exec snapd from core snap
+        apt install -y --allow-downgrades snapd=2.55.3+22.04
+    fi
+    snap install core --revision=17200 # 16-2.61.4-20240607
+
+    # installation of core should not trigger prerequisites, but let's double check snapd is not installed
+    ! snap list snapd
```
Comment (sergiocazzolato) [inline tests/main/snapd-snap-transition-auto-install/task.yaml:43]: perhaps: not snap list snapd

**Review by sergiocazzolato — APPROVED** [review]
some questions inline

**Review by andrewphelpsj — APPROVED** [review]
Thanks, looks good to me! Removing the experimental feature is a good idea, yes.

File: tests/main/snapd-snap-transition-auto-install/task.yaml (hunk 9 above)
Comment (ernestl) [reply tests/main/snapd-snap-transition-auto-install/task.yaml:22]: added check

File: tests/main/snapd-snap-transition-auto-install/task.yaml (hunk 11 above)
Comment (ernestl) [reply tests/main/snapd-snap-transition-auto-install/task.yaml]: changed

File: tests/main/snapd-snap-transition-auto-install/task.yaml (hunk 10 above)
Comment (ernestl) [reply tests/main/snapd-snap-transition-auto-install/task.yaml:23]: added check

File: tests/main/snapd-snap-transition-auto-install/task.yaml (hunk 8 above)
Comment (ernestl) [reply tests/main/snapd-snap-transition-auto-install/task.yaml:112]: Prefer to check from a clean reference, regardless of previous activity.

**Comment by ernestl** [comment]
Failures:

fedora-os |
- openstack:fedora-40-64:tests/main/selinux-clean | known, unrelated

**Pull request merged**